### PR TITLE
Linting README.md and all_documents.md

### PR DESCRIPTION
**Pull request by ns-rse**
Introduces [pre-commit](https://pre-commit.com) to link Markdown with [markdownlint-cli2](https://github.com/DavidAnson/markdownlint-cli2) and a few other hooks from `pre-commit` (`trailing-whitespace`, `end-of-file-fixer` and `check-yaml`).

* Line length is set to 120 for `markdownlint-cli2` but ignored for code-blocks and tables. 
* Files with the extension `.md` and `.markdown` are linted. 
* Issues that can be fixed automatically are.

`.pre-commit-config.yaml` includes a section to enable [`pre-commit.ci`](https://pre-commit.ci) but this requires linking the organisation to to `pre-commit.ci` and enabling for the repository (I've written a blog post on this the relevant section is [here](https://ns-rse.github.io/posts/pre-commit-ci/#setup)).

I can work through linting the existing code base in a separate PR but thought it would be useful to get feedback on the current selection of rules and how happy people are with this before forging ahead.

### CODE REVIEW

**Review by davehorsfall — COMMENTED** [review]
@ns-rse thanks very much for this PR. 

I haven't used `pre-commit` before. I use linters through my IDE (vscode) were changes are applied automatically on save prior to commit. However, it is good to explore this, and happy to include these pre-commit configs and adopt this practice for the repo. It has been a bit of a learning curve, and I had a question. 

If try to commit a file that fails the check for the `end-of-file-fixer` hook, a fix is applied automatically to the file. I can add and commit the modified file as expected. However, if I try to commit a markdown file that fails the check for the `markdownlint-cli2` hook, the output will tell me the test has failed and it will show me all the errors with the file (and errors for all other markdown files in the repo), but it won't apply fixes. 

```
(base) ndh122@xxx:~/Projects/RSEToolkit/rse-competencies-toolkit$ git commit -m "testing pre-commit linting"
trim trailing whitespace.............................(no files to check)Skipped
fix end of files.........................................................Passed
check yaml...........................................(no files to check)Skipped
markdownlint-cli2........................................................Failed
- hook id: markdownlint-cli2
- exit code: 1
- files were modified by this hook

markdownlint-cli2 v0.10.0 (markdownlint v0.31.1)
Finding: about.md **/*.md **/*.markdown
Linting: 8 file(s)
Summary: 113 error(s)
_site/all_documents.md:9:121 MD013/line-length Line length [Expected: 120; Actual: 153]
_site/all_documents.md:11:121 MD013/line-length Line length [Expected: 120; Actual: 166]
_site/LICENSE.md:1 MD041/first-line-heading/first-line-h1 First line in a file should be a top-level heading [Context: "Unless otherwise noted, all co..."]
_site/LICENSE.md:5:121 MD013/line-length Line length [Expected: 120; Actual: 142]
_site/LICENSE.md:9:121 MD013/line-length Line length [Expected: 120; Actual: 205]
_site/LICENSE.md:11:121 MD013/line-length Line length [Expected: 120; Actual: 190]
_site/LICENSE.md:13:121 MD013/line-length Line length [Expected: 120; Actual: 755]
_site/LICENSE.md:15:121 MD013/line-length Line length [Expected: 120; Actual: 175]
_site/LICENSE.md:18:121 MD013/line-length Line length [Expected: 120; Actual: 541]
_site/LICENSE.md:22:121 MD013/line-length Line length [Expected: 120; Actual: 557]
_site/LICENSE.md:23:121 MD013/line-length Line length [Expected: 120; Actual: 192]
_site/LICENSE.md:24:121 MD013/line-length Line length [Expected: 120; Actual: 380]
_site/LICENSE.md:25:121 MD013/line-length Line length [Expected: 120; Actual: 270]
_site/LICENSE.md:26:121 MD013/line-length Line length [Expected: 120; Actual: 180]
_site/LICENSE.md:27:121 MD013/line-length Line length [Expected: 120; Actual: 133]
_site/LICENSE.md:28:121 MD013/line-length Line length [Expected: 120; Actual: 257]
_site/LICENSE.md:30:121 MD013/line-length Line length [Expected: 120; Actual: 409]
_site/LICENSE.md:31:121 MD013/line-length Line length [Expected: 120; Actual: 294]
_site/LICENSE.md:36:121 MD013/line-length Line length [Expected: 120; Actual: 232]
_site/LICENSE.md:39:121 MD013/line-length Line length [Expected: 120; Actual: 204]
_site/LICENSE.md:41:121 MD013/line-length Line length [Expected: 120; Actual: 621]
_site/LICENSE.md:43:121 MD013/line-length Line length [Expected: 120; Actual: 222]
_site/LICENSE.md:44:121 MD013/line-length Line length [Expected: 120; Actual: 276]
_site/LICENSE.md:45:121 MD013/line-length Line length [Expected: 120; Actual: 337]
_site/LICENSE.md:48:121 MD013/line-length Line length [Expected: 120; Actual: 373]
_site/LICENSE.md:50:121 MD013/line-length Line length [Expected: 120; Actual: 335]
_site/LICENSE.md:60:121 MD013/line-length Line length [Expected: 120; Actual: 198]
_site/LICENSE.md:66:121 MD013/line-length Line length [Expected: 120; Actual: 143]
_site/LICENSE.md:67:121 MD013/line-length Line length [Expected: 120; Actual: 302]
_site/LICENSE.md:68:121 MD013/line-length Line length [Expected: 120; Actual: 137]
_site/LICENSE.md:69:121 MD013/line-length Line length [Expected: 120; Actual: 167]
_site/LICENSE.md:74:121 MD013/line-length Line length [Expected: 120; Actual: 166]
_site/LICENSE.md:75:121 MD013/line-length Line length [Expected: 120; Actual: 252]
_site/LICENSE.md:77:121 MD013/line-length Line length [Expected: 120; Actual: 188]
_site/LICENSE.md:80:121 MD013/line-length Line length [Expected: 120; Actual: 634]
_site/LICENSE.md:81:121 MD013/line-length Line length [Expected: 120; Actual: 550]
_site/LICENSE.md:82:121 MD013/line-length Line length [Expected: 120; Actual: 212]
_site/LICENSE.md:85:121 MD013/line-length Line length [Expected: 120; Actual: 216]
_site/LICENSE.md:88:121 MD013/line-length Line length [Expected: 120; Actual: 127]
_site/LICENSE.md:90:121 MD013/line-length Line length [Expected: 120; Actual: 154]
_site/LICENSE.md:91:121 MD013/line-length Line length [Expected: 120; Actual: 226]
_site/LICENSE.md:96:121 MD013/line-length Line length [Expected: 120; Actual: 183]
_site/LICENSE.md:99:121 MD013/line-length Line length [Expected: 120; Actual: 246]
_site/LICENSE.md:100:121 MD013/line-length Line length [Expected: 120; Actual: 337]
_site/LICENSE.md:101:121 MD013/line-length Line length [Expected: 120; Actual: 140]
_site/LICENSE.md:102:121 MD013/line-length Line length [Expected: 120; Actual: 234]
_site/README.md:13:121 MD013/line-length Line length [Expected: 120; Actual: 235]
_site/README.md:15:121 MD013/line-length Line length [Expected: 120; Actual: 256]
_site/README.md:19:121 MD013/line-length Line length [Expected: 120; Actual: 257]
_site/README.md:26:121 MD013/line-length Line length [Expected: 120; Actual: 210]
_site/README.md:28:121 MD013/line-length Line length [Expected: 120; Actual: 152]
_site/README.md:46:121 MD013/line-length Line length [Expected: 120; Actual: 234]
_site/README.md:50:121 MD013/line-length Line length [Expected: 120; Actual: 203]
_site/README.md:54:121 MD013/line-length Line length [Expected: 120; Actual: 190]
_site/README.md:68:121 MD013/line-length Line length [Expected: 120; Actual: 172]
_site/README.md:82:121 MD013/line-length Line length [Expected: 120; Actual: 158]
_site/README.md:143:121 MD013/line-length Line length [Expected: 120; Actual: 167]
_site/README.md:149 MD040/fenced-code-language Fenced code blocks should have a language specified [Context: "```"]
_site/README.md:160:121 MD013/line-length Line length [Expected: 120; Actual: 220]
about.md:6:121 MD013/line-length Line length [Expected: 120; Actual: 235]
about.md:8:121 MD013/line-length Line length [Expected: 120; Actual: 256]
about.md:12:121 MD013/line-length Line length [Expected: 120; Actual: 256]
about.md:19:121 MD013/line-length Line length [Expected: 120; Actual: 209]
about.md:21:121 MD013/line-length Line length [Expected: 120; Actual: 151]
about.md:31:121 MD013/line-length Line length [Expected: 120; Actual: 321]
about.md:35:121 MD013/line-length Line length [Expected: 120; Actual: 152]
about.md:37 MD036/no-emphasis-as-heading/no-emphasis-as-header Emphasis used instead of a heading [Context: "Active collaborative documents"]
about.md:42 MD036/no-emphasis-as-heading/no-emphasis-as-header Emphasis used instead of a heading [Context: "Working documents for the fram..."]
about.md:49 MD036/no-emphasis-as-heading/no-emphasis-as-header Emphasis used instead of a heading [Context: "Notes from community sessions ..."]
LICENSE.md:1 MD041/first-line-heading/first-line-h1 First line in a file should be a top-level heading [Context: "Unless otherwise noted, all co..."]
LICENSE.md:5:121 MD013/line-length Line length [Expected: 120; Actual: 142]
LICENSE.md:9:121 MD013/line-length Line length [Expected: 120; Actual: 205]
LICENSE.md:11:121 MD013/line-length Line length [Expected: 120; Actual: 190]
LICENSE.md:13:121 MD013/line-length Line length [Expected: 120; Actual: 755]
LICENSE.md:15:121 MD013/line-length Line length [Expected: 120; Actual: 175]
LICENSE.md:18:121 MD013/line-length Line length [Expected: 120; Actual: 541]
LICENSE.md:22:121 MD013/line-length Line length [Expected: 120; Actual: 557]
LICENSE.md:23:121 MD013/line-length Line length [Expected: 120; Actual: 192]
LICENSE.md:24:121 MD013/line-length Line length [Expected: 120; Actual: 380]
LICENSE.md:25:121 MD013/line-length Line length [Expected: 120; Actual: 270]
LICENSE.md:26:121 MD013/line-length Line length [Expected: 120; Actual: 180]
LICENSE.md:27:121 MD013/line-length Line length [Expected: 120; Actual: 133]
LICENSE.md:28:121 MD013/line-length Line length [Expected: 120; Actual: 257]
LICENSE.md:30:121 MD013/line-length Line length [Expected: 120; Actual: 409]
LICENSE.md:31:121 MD013/line-length Line length [Expected: 120; Actual: 294]
LICENSE.md:36:121 MD013/line-length Line length [Expected: 120; Actual: 232]
LICENSE.md:39:121 MD013/line-length Line length [Expected: 120; Actual: 204]
LICENSE.md:41:121 MD013/line-length Line length [Expected: 120; Actual: 621]
LICENSE.md:43:121 MD013/line-length Line length [Expected: 120; Actual: 222]
LICENSE.md:44:121 MD013/line-length Line length [Expected: 120; Actual: 276]
LICENSE.md:45:121 MD013/line-length Line length [Expected: 120; Actual: 337]
LICENSE.md:48:121 MD013/line-length Line length [Expected: 120; Actual: 373]
LICENSE.md:50:121 MD013/line-length Line length [Expected: 120; Actual: 335]
LICENSE.md:60:121 MD013/line-length Line length [Expected: 120; Actual: 198]
LICENSE.md:66:121 MD013/line-length Line length [Expected: 120; Actual: 143]
LICENSE.md:67:121 MD013/line-length Line length [Expected: 120; Actual: 302]
LICENSE.md:68:121 MD013/line-length Line length [Expected: 120; Actual: 137]
LICENSE.md:69:121 MD013/line-length Line length [Expected: 120; Actual: 167]
LICENSE.md:74:121 MD013/line-length Line length [Expected: 120; Actual: 166]
LICENSE.md:75:121 MD013/line-length Line length [Expected: 120; Actual: 252]
LICENSE.md:77:121 MD013/line-length Line length [Expected: 120; Actual: 188]
LICENSE.md:80:121 MD013/line-length Line length [Expected: 120; Actual: 634]
LICENSE.md:81:121 MD013/line-length Line length [Expected: 120; Actual: 550]
LICENSE.md:82:121 MD013/line-length Line length [Expected: 120; Actual: 212]
LICENSE.md:85:121 MD013/line-length Line length [Expected: 120; Actual: 216]
LICENSE.md:88:121 MD013/line-length Line length [Expected: 120; Actual: 127]
LICENSE.md:90:121 MD013/line-length Line length [Expected: 120; Actual: 154]
LICENSE.md:91:121 MD013/line-length Line length [Expected: 120; Actual: 226]
LICENSE.md:96:121 MD013/line-length Line length [Expected: 120; Actual: 183]
LICENSE.md:99:121 MD013/line-length Line length [Expected: 120; Actual: 246]
LICENSE.md:100:121 MD013/line-length Line length [Expected: 120; Actual: 337]
LICENSE.md:101:121 MD013/line-length Line length [Expected: 120; Actual: 140]
LICENSE.md:102:121 MD013/line-length Line length [Expected: 120; Actual: 234]
```

My understanding is that fixes would be applied automatically. Should they? Am I doing something stupid? Thought I would ask you first, rather than battle with this. Thanks!

File: README.md [hunk 1]
```diff
@@ -10,62 +10,76 @@ The idea is to construct a resource that is for the RSE community and curated by
...
 
 ### Current Status
 
 - [x] Initial Research
-- [ ] Initial Development <-- We are Here  
+- [ ] Initial Development <-- We are Here
 - [ ] Minimum viable product
-- [ ] Alpha Release  
-- [ ] Feature-Complete Release  
+- [ ] Alpha Release
```
Comment (anenadic) [inline README.md:42]: Do we need a beta release before the alpha and the feature-complete as an intermediate step? Be more iterative?

**Review by anenadic — COMMENTED** [review]
I agree that this PR a very good idea @ns-rse  - I do not have enough knowledge of these hooks to comment on the code. Left a comment on the proposed stages of development though to make it more iterative. Also would be a good idea to think about how we are going to pilot/evaluate this in the earlier stages so we have some direction on how we will refine this.

**Comment by ns-rse** [comment]
Hi @davehorsfall @anenadic,

Thanks for taking the time to look through this PR.

To answer @davehorsfall question, you're not doing anything stupid at all.

Some fixes should be fixed automatically by `markdownlint-cli2` and other hooks where its possible to do so. As with Git merge conflicts it can't always know the right thing to do, or in the case of `MD012/line-length` wrapping lines to the suggested length (in this case `120`) hasn't been implemented in `markdownlint-cli2`. Another example that the linter has picked up is `MD036/no-emphasis-as-heading/no-emphasis-as-header` and it can't correct this as it doesn't know what level of heading it should be.

As you've seen in such instances we just see a report of what needs fixing.

To keep this PR small as its introducing a new way of linting I figured I'd just focus on a couple of files first to make the PR and the changes its introducing easier to review and I'm happy to go through and lint all the existing code in a separate PR otherwise whenever a file gets touched and committed again people will encounter errors in code they haven't touched (which reminds me I should really have introduced a [`.git-blame-ignore-revs`](https://rse.shef.ac.uk/blog/git-blame/) so that blame is not incorrectly attributed.

In the longer term enabling the `pre-commit.ci` would I think be sensible as it catches instances where contributors haven't enabled `pre-commit` locally and fixed all issues prior to committing.


More than happy to add @anenadic suggestion of a Beta Release to this PR.

**Comment by davehorsfall** [comment]
Thanks @ns-rse for the clarification. Happy to merge this now, and then welcome another PR with subsequent fixes for remaining code.

**Review by davehorsfall — APPROVED** [review]

**Comment by ns-rse** [comment]
Great, thanks @davehorsfall :+1: 

I'll get onto that over the coming days and also add a `.git-blame-ignore-revs` file so that blame shows who wrote rather than linted the code.

**Pull request merged**